feat(runner): add support for command delay and parallel setup

hargut opened the pull request:
Following the discussion in #242 I've started to create the API and draft implementation for the Delay support.
This is still a WIP, but in case you can provide some early feedback I'd be happy to incorporate that.

The works for setup_in_parallel are not finished and in progress. The Delay area is quite complete.

gamma0987 commented:
Thanks for working on this. I like that you're testing the delay execution in unit tests. You can use `rstest` (it's already in the dev-dependencies) if you haven't noticed yet.

I hope it helps, here some impressions:

The `iai_callgrind_runner::api` is an adapter/layer between `iai-callgrind` and the final logic in the `iai-callgrind-runner`, so they can communicate. You can orient yourself by the other structs in the api. For example see `Command` and how it is implemented from the `iai-callgrind` library through the api to the final `Command` in `iai_callgrind_runner::runner::bin_bench`. The structures in the api store merely configurations used by the final component implemented somewhere else in the iai-callgrind-runner. The final component (in your case `iai_callgrind_runner::runner::bin_bench::Delay`) applies the defaults, not the Delay in the api. The api is especially the wrong place for any logic like `exec_delay_fn`. This function should be part of the `iai_callgrind_runner::runner::bin_bench::Delay`.

Please use the builder pattern for the user-side structs in the `iai-callgrind` library. 

I would prefer the name `DelayKind` over `ExecutionDelay` because that's the naming scheme already in use.

Why so many `from_*` methods? Wouldn't be `impl From<???> for iai_callgrind::bin_bench::Delay` more convenient from a user perspective?

hargut commented:
Thank you for the feedback. I'll try to adapt accordingly.

> Why so many from_* methods? Wouldn't be impl From<???> for iai_callgrind::bin_bench::Delay more convenient from a user perspective?


Is that possible to be implemented for various possible options with the same generic type requirements?
Maybe I'm missing something or I'm not aware of some Rust possibilities.

For example in the current layout there are:
```rust
pub fn from_path<T: Into<PathBuf>>(path: T) -> Self {}
pub fn from_unix_socket<T: Into<PathBuf>>(path: T) -> Self {}
```
Both are built around  the generic type `Into<PathBuf>` but being a different kind. How can such scenario be fit into a `From` implementation? 
```rust
impl<T> From<T> for iai_callgrind::bin_bench::Delay where T: Into<PathBuf> {
    fn from(value: T) -> Self {
        todo!()
    }
}
```
Currently it would be possible for the `from_duration()` as there are no overlaps, but I'm not sure if:
`Delay::from(Duration::from_secs(60))` would be more convenient or clearer for the user than `Delay::from_secs(60)`.

Offering `Command::new("/bin/echo").delay(Duration::from_secs(60).into())` could also be beneficial.

For the first draft, I've selected the multiple `from_*` methods, and tried to be as generic/clear as possible.
But I was not sure if this meets the expectations.


Kind regards,
Harald

gamma0987 commented:
> Is that possible to be implemented for various possible options with the same generic type requirements?

Not that I know of.

But you could easily implement `impl From<Duration> for Delay` instead of `Delay::from_duration(...)` and the others `from_millis`, `from_secs` and then use it in `Command::delay(Duration::from_micros(...))` if you would use `fn delay<T>(delay: T) where T: Into<Delay> { ... }` as function signature.

Is the `DelayKind::UnixSocket` actually required? Wouldn't it be enough to check if the path exist with `DelayKind::File`?

gamma0987 commented:
I'll have a closer look at it soon. From a quick glance, `iai_callgrind::bin_bench::Delay` needs a complete rewrite. Please have a look at the other builders in the same file and apply the pattern you see there. The `build` method is redundant. `new` should take a DelayKind as argument. The defaults for poll time, timeout are applied in the runner not in `iai-callgrind`. 

Why is DelayKind in `iai-callgrind` a struct and not an enum? You might use the enum from the api directly.

hargut commented:
Thank you very much for your feedback. :+1: 

I've refactored `iai_callgrind::bin_bench::Delay`
  - `Delay::new` now takes a `DelayKind` as argument.
  - setting the defaults is moved to `iai-callgrind-runner::runner::bin_bench`
  - What did you mean with?
    > The `build` method is redundant.

    At the moment it is the same as with the `Command` and I think the intention is to be able to re-use the builder. Isn't the `build()` method also a characteristic of the builder pattern?


> Why is DelayKind in `iai-callgrind` a struct and not an enum? You might use the enum from the api directly.

It was an error that this line was present in the code as it was not used anywhere.
Initially my approach was to not expose `DelayKind` at all (only using it internally in `iai-callgrind::bin_bench`) and use `from*()` for creation of the `Delay` and omit the `build()`. From my point of view the `DelayKind` complicates the situation for the user.

Currently both approaches are present and part of the examples:
```rust
let duration = Command::new(env!("echo")).delay(
     Delay::new(DelayKind::DurationElapse(Duration::from_secs(60))).build()
);
let duration_from = Command::new(env!("echo")) .delay(
     Delay::from(Duration::from_secs(60))
);

let path = Command::new("echo").delay(
     Delay::new(DelayKind::PathExists("/your/path/to/wait/for".into())).build()
);
let path_from =  Command::new("echo").delay(
     Delay::from_path("/your/path/to/wait/for")
);

let tcp = Command::new("echo").delay(
     Delay::new(DelayKind::TcpConnect("127.0.0.1:31000".parse::<SocketAddr>().unwrap())).build(),
);
let tcp_from = Command::new("echo").delay(
     Delay::from_tcp_socket("127.0.0.1:31000".parse::<SocketAddr>().unwrap()),
);
```

Kind regards,
Harald

hargut commented:
The pipeline error seems to be related to:
```
2024-09-05T09:17:49.8090086Z   = note: /usr/bin/ld: final link failed: No space left on device
2024-09-05T09:17:49.8090606Z           collect2: error: ld returned 1 exit status
```

gamma0987 commented:
Thanks for changing the `Delay` so far. 

> At the moment it is the same as with the Command and I think the intention is to be able to re-use the builder.

`Command` is the only builder in `iai-callgrind` with a `build` method because it was unavoidable and had advantages. You can re-use all builders in iai-callgrind without the build method. Adding this method just makes people think it is required.

> Isn't the build() method also a characteristic of the builder pattern?

Not necessarily a `build` method but usually something like that. But, it's not needed for the builders in iai-callgrind and it is just confusing. `command.delay(Delay::new(...))` or `comand.delay(Delay::new(...).timeout(...))` is fully sufficient.

`udp` in addition to `tcp` should be covered, too.

hargut commented:
Removed the `build()` method from the `Delay`.

For `UDP` I'm not really sure if it is good to offer that functionality. UDP itself does not have a `Connection` concept within the protocol.
The only reliable way to detect `Connected` is by sending a request and checking the response. That being said, this requires knowledge of the listener connecting to in order to decide if it was successful.

It could be possible to do that but would require as a minimum something like:
```rust
DelayKind::UdpConnect(SocketAddr, Request(Vec<u8>), Response(Vec<u8>))
```
This would as well require to deal with timeouts, resending data and client related tasks. Thinking of protocols with handshakes (e.g. QUIC) or other binary/encrypted scenarios this could get complex.

gamma0987 commented:
Couldn't this be done on a best effort basis? I think, if someone is using udp and wants to wait for the udp connection, he knows about his server and what to send so that the response is not empty if the connection is ready. Maybe it's not even necessary to know about the exact response and checking the received number of bytes are not zero is sufficient. Otherwise, he can still use one of the other delay methods.

hargut commented:
Hi @gamma0987,

I've added the `DelayKind::UdpResponse` which checks if a single byte is part of the response.

Kind regards,
Harald

hargut commented:
Added a small update, that removes the `local_addr` from the `DelayKind::UdpResponse`.
The change now behaves similiar to `DelayKind::TcpConnect` for the local socket creation.

gamma0987 commented:
Is this ready for review? If yes, just hit the button or ping me.

hargut commented:
Hi @gamma0987,

yes, it's ready for review. 😃

Kind regards,
Harald

gamma0987 reviewed (changes requested):
Thanks for working on this!

hargut commented on iai-callgrind-runner/src/runner/common.rs:
```diff
@@ -138,6 +148,20 @@ impl Assistant {
         }
 
+        if self.run_parallel {
+            let command = match nocapture {
```
This is a mistake and looks like I wanted to come back to that area to finish the pipe setups.
The intention is to keep the same behaviour for the `Stdio` config as with regular runs.

The `std::process::Command` execution methods have different default setups:
- [Command.output()](https://doc.rust-lang.org/std/process/struct.Command.html#method.output) By default, stdout and stderr are captured
- [Command.status()](https://doc.rust-lang.org/std/process/struct.Command.html#method.status) By default, stdin, stdout and stderr are inherited from the parent.
- [Command.spawn()](https://doc.rust-lang.org/std/process/struct.Command.html#method.spawn) By default, stdin, stdout and stderr are inherited from the parent.

To have the same behaviour as with the default command runs changed it to keep pipe setups from `nocapture.apply(&mut command)`,
and setting `command.stdin(Stdio::null())` for the `NoCapture::False` case as this is executed with `Command.output()` in regular runs.

hargut commented on iai-callgrind/src/bin_bench.rs:
```diff
@@ -181,6 +183,164 @@ pub struct BinaryBenchmark {
+pub use internal::InternalDelayKind as DelayKind;
+
+impl Delay {
+    /// Instantiate a [`Delay`] which will wait until a path exists ([`Path::exists`]).
```
Added examples to the methods.

hargut commented on iai-callgrind-runner/src/runner/bin_bench.rs:
```diff
@@ -873,3 +900,343 @@ impl Benchmark for SaveBaselineBenchmark {
+            };
+        } else {
+            match rx.recv() {
+                Ok(_) => handle.join().unwrap(),
```
Converted the `unwrap()`s to Errors/Results and propagated them.

hargut commented on iai-callgrind/src/bin_bench.rs:
```diff
@@ -1649,6 +1809,91 @@ impl Command {
+    /// use iai_callgrind::{binary_benchmark_group, binary_benchmark, main};
+    ///
+    /// fn setup_server() {
+    ///     todo!()
```
Updated the example and remove the `todo!()`.

hargut commented on iai-callgrind/src/bin_bench.rs:
```diff
@@ -1649,6 +1809,91 @@ impl Command {
+    /// Supported events are:
+    ///  - Timer expired
+    ///  - File path exists
+    ///  - TCP connect succeeded
```
Applied.

hargut commented on iai-callgrind/src/bin_bench.rs:
```diff
@@ -181,6 +183,164 @@ pub struct BinaryBenchmark {
+
+    /// Instantiate a [`Delay`] waiting until an event has happened.
+    ///
+    /// The possible events are defined in [`DelayKind`]. The following kinds are supported:
```
Removed the list, only kept the reference to `DelayKind`.

hargut commented on iai-callgrind-runner/src/runner/bin_bench.rs:
```diff
@@ -873,3 +900,343 @@ impl Benchmark for SaveBaselineBenchmark {
+        match self.kind {
+            DelayKind::DurationElapse(_) => {
+                if self.poll.is_some() {
+                    warn!("Ignoring poll setting. Not support for {:?}", self.kind);
```
Corrected.

hargut commented on iai-callgrind-runner/src/runner/bin_bench.rs:
```diff
@@ -873,3 +900,343 @@ impl Benchmark for SaveBaselineBenchmark {
+            }
+        }
+        if let (Some(poll), Some(timeout)) = (self.poll, self.timeout) {
+            if poll >= timeout {
```
Changed the behaviour to set `poll` to `timeout -5ms` in case `timeout >= 5ms`, otherwise using `timeout`.
Updated the documentation for `Command::poll` & `Command::timeout`.

hargut commented:
Hi @gamma0987,

I've addressed the points from your review and updated the branch to include the changes.

Thank you for your feedback. :+1: 

Kind regards,
Harald

gamma0987 commented:
Hi. Thanks for your initial work on this. There are still some things wrong and missing but I think I can take it as a base the way it is. Your original commit will of course be preserved.

hargut commented:
Hi @gamma0987,

thank you for finalizing the works & fixing the open issues. 👍 

Kind regards,
Harald
